### PR TITLE
Port Tensor1D to new framework

### PR DESCRIPTION
**Pull request by bollu**
We see the dreaded error

> -- Error: unknown free variable: _kernel_fresh.459

in our instance definition

```lean
instance : OpDenote ExOp ExTy where
...
```

This suggests that we should spend the time working around
or fixing this kernel bug.

### CODE REVIEW

**Comment by bollu** [comment]
Ping @tobiasgrosser .

**Comment by tobiasgrosser** [comment]
@bollu, can this be rebased and merged?

File: SSA/Projects/Tensor1D/Tensor1D.lean [hunk 1]
```diff
@@ -294,142 +294,74 @@ theorem Tensor1d.tile [Inhabited α] (t : Tensor1d α) (SIZE : 4 ∣ t.size) (f
...
-
-
-
 namespace ArithScfLinalg
 
-open Val
 
-inductive Op
+
+inductive ExOp
```
Comment (goens) [inline SSA/Projects/Tensor1D/Tensor1D.lean:301]: why the rename?

File: SSA/Projects/Tensor1D/Tensor1D.lean [hunk 2]
```diff
@@ -294,142 +294,74 @@ theorem Tensor1d.tile [Inhabited α] (t : Tensor1d α) (SIZE : 4 ∣ t.size) (f
...
-    let y : Int := y;
-    x + y
-  | .sub =>
-    let (x, y) := arg;
+
+instance : OpSignature ExOp ExTy where
+  outTy : ExOp → ExTy
+  | .add => .int
+  | .sub => .int
+  | .const _ => .nat
```
Comment (goens) [inline SSA/Projects/Tensor1D/Tensor1D.lean:325]: why does `.const` go to `.nat` and not to `.int`?

File: SSA/Projects/Tensor1D/Tensor1D.lean [hunk 3]
```diff
@@ -294,142 +294,74 @@ theorem Tensor1d.tile [Inhabited α] (t : Tensor1d α) (SIZE : 4 ∣ t.size) (f
...
+@[reducible]
+instance : OpDenote ExOp ExTy where
+  denote
+  | .const v, _, _ => v
+  | .add, (.cons x (.cons y nil)), _ =>
+      let x : Int := x;
+      let y : Int := y
+      x + y
+  | .sub, (.cons x (.cons y nil)), _ =>
     let x : Int := x;
```
Comment (goens) [inline SSA/Projects/Tensor1D/Tensor1D.lean:350]: does this add implicit coercions, or force `x` and `y` to be `.int`s? why don't you use the `toTuple` function? I guess it's a style question but I'd find it more readable

**Review by goens — CHANGES_REQUESTED** [review]
I find it strange having different `.int` and `.nat` base types and how you can only build constants of `.nat` type - how are you supposed to convert to int? Is it really necessary to have both (couldn't you just have one and either have `0 - n = 0` like in `Nat` or `extract t n m` return garbage for negative `n` or `m`?

Actionably: can I at least please have a docstring documenting the ops and what they do and the relationship between ints and nats here?

File: SSA/Projects/Tensor1D/Tensor1D.lean (hunk 3 above)
Comment (bollu) [reply SSA/Projects/Tensor1D/Tensor1D.lean:350]: It forces `x` and `y` to be `.int`. I played around with other ways of doing it, including `toTuple`, but I was unable to convince `Lean` that `Int` was defeq to `[[ExTy.int]]` without this kind of explicit let-shadowing.

File: SSA/Projects/Tensor1D/Tensor1D.lean (hunk 2 above)
Comment (tobiasgrosser) [reply SSA/Projects/Tensor1D/Tensor1D.lean]: Do you you use `.add` for index computation as well? I am surprised add/sub/const seem to belong together, so either are value computation of index computation. Maybe name them index_add, index_sub, index_const as well as .add, .sub, .const for the value domain?

**Comment by bollu** [comment]
@goens addressed comments!

File: SSA/Projects/Tensor1D/Tensor1D.lean (hunk 1 above)
Comment (bollu) [reply SSA/Projects/Tensor1D/Tensor1D.lean:313]: To be consistent with the others in InstCombine

File: SSA/Projects/Tensor1D/Tensor1D.lean (hunk 2 above)
Comment (bollu) [reply SSA/Projects/Tensor1D/Tensor1D.lean]: @tobiasgrosser The IR is limited, I only use add/sub for values. Happy to extend the name.

File: SSA/Projects/Tensor1D/Tensor1D.lean (hunk 3 above)
Comment (bollu) [reply SSA/Projects/Tensor1D/Tensor1D.lean:362]: feel invited to give it a go!

File: SSA/Projects/Tensor1D/Tensor1D.lean (hunk 2 above)
Comment (bollu) [reply SSA/Projects/Tensor1D/Tensor1D.lean]: Done, remaned.

File: SSA/Projects/Tensor1D/Tensor1D.lean (hunk 1 above)
Comment (goens) [reply SSA/Projects/Tensor1D/Tensor1D.lean:313]: Not sure I understand, `InstCombine` doesn't call them `ExOp`, it calls them `Op` (and similarly for `Ty`)

File: SSA/Core/ErasedContext.lean [hunk 4]
```diff
@@ -198,6 +199,11 @@ theorem Valuation.snoc_toSnoc {Γ : Ctxt Ty} {t t' : Ty} (s : Γ.Valuation) (x :
     (v : Γ.Var t') : (s.snoc x) v.toSnoc = s v := by
   simp [Ctxt.Valuation.snoc]
 
+/-- Build valuation from a list of values of types `types`. -/
+def Valuation.ofList {types : List Ty} : HVector toType types → Valuation (Ctxt.ofList types)
```
Comment (alexkeizer) [inline SSA/Core/ErasedContext.lean:203]: I feel like this should be `Valuation.ofHVector`, but I guess arguments can be made for both

File: SSA/Projects/Tensor1D/Tensor1D.lean (hunk 1 above)
Comment (alexkeizer) [reply SSA/Projects/Tensor1D/Tensor1D.lean:313]: I'd argue for sticking with `Op` as well.
I think @bollu is referring to the example dialects in the `Framework.lean` (formerly `IntrinsicAsymptotics`) file, which do use  the`Ex-` prefix to make it explicit they are examples, but those should arguably be put in their own file anyway

File: SSA/Projects/Tensor2D/Tensor2D.lean [hunk 5]
```diff
@@ -205,14 +168,14 @@ theorem map_fill_2d
...
-  (e := e) (Op := Op) (β := BaseType) [dsl_bb2|
+  (e := e) (Op := Op) (β := ) [dsl_bb2|
     return op:fill2d (op:constInt(i) (), 
       op:extract2d (
           ((op:constIx(sz₀) () , op:constIx(sz₁) ()),
            (op:constIx(ix₀) (), op:constIx(ix₁) ())),
            op:constTensor(t) ()))
   ] =
-  TSSA.eval (e := e) (Op := Op) (β := BaseType) [dsl_bb2|
+  TSSA.eval (e := e) (Op := Op) (β := ) [dsl_bb2|
```
Comment (alexkeizer) [inline SSA/Projects/Tensor2D/Tensor2D.lean:178]:
```suggestion
  TSSA.eval (e := e) (Op := Op) [dsl_bb2|
```
I'm surprised this is legal syntax, but we probably want to remove the unspecified argument completely (also at line 171 above).

**Review by alexkeizer — COMMENTED** [review]
Looks good, besides some minor nitpicks

File: SSA/Core/ErasedContext.lean (hunk 4 above)
Comment (bollu) [reply SSA/Core/ErasedContext.lean]: Renamed as requested

File: SSA/Projects/Tensor2D/Tensor2D.lean (hunk 5 above)
Comment (alexkeizer) [reply SSA/Projects/Tensor2D/Tensor2D.lean]: You forgot the one on line 171, so I took the liberty of removing it

**Review by alexkeizer — APPROVED** [review]
LGTM, @goens feel free to merge if you think your concerns have been addressed

**Pull request merged**